File: mcp/src/hooks/userprompt-submit.ts
/**
 * Hook: UserPromptSubmit — Layer 1 of the v0.4.0 self-ping enforcement layer.
 *
 * On every user prompt mneme injects a reminder block that:
 *   1. Lists the top 3 mneme tools most relevant to the user's message.
 *   2. Shows a "trespass log" of Grep/Read calls in this session that did
 *      NOT go through mneme first — training the AI to notice the pattern.
 *
 * The existing `inject.ts` hook (Mode B smart inject) continues to run and
 * injects semantic context from the shard. This hook is ADDITIVE — it runs
 * in parallel with inject.ts when Claude Code routes UserPromptSubmit. Both
 * outputs arrive as `additionalContext` prepended to the turn.
 *
 * Hook output protocol (Claude Code spec):
 *   { "hook_specific": { "additionalContext": "<block>" } }
 *   Exit 0 = continue. This hook never blocks.
 *
 * Fail-open guarantee: any error → return empty additionalContext, exit 0.
 * mneme MUST NOT interrupt the user's workflow.
 */

import { getSessionTrespasses } from "./lib/recent-tool-calls.ts";
import { getHooksConfig } from "./lib/recent-tool-calls.ts";
import { errMsg } from "../errors.ts";

// ---------------------------------------------------------------------------
// Item #119 (2026-05-05) — smart context injection.
//
// The pre-fix hook fired the FULL reminder block (~200 tokens) on
// EVERY UserPromptSubmit, including simple acks ("ok", "continue",
// "thanks"). Across a 50-turn session that's ~10K tokens of pure
// overhead. Bench harness confirmed mneme was NOT saving tokens net
// (1.34× measured vs CRG's 6.8× claimed) — partly because the
// reminder added more than the tool savings recouped.
//
// Three-tier classification:
//   - "simple"  — small-talk / acks / planning chat. Inject nothing.
//   - "code"    — code-related question. Inject the LIGHT block:
//                 top-3 tools only, no trespass log unless real
//                 trespasses exist. ~100-150 tokens.
//   - "resume"  — continuation cue ("continue", "where was i", etc.)
//                 with the implicit assumption that a compact may
//                 have wiped earlier context. Inject the HEAVY
//                 block: top-3 tools + trespass log + cue to call
//                 mneme_resume / step_status. ~400-500 tokens.
//
// Distribution across a real session (eyeballed): ~30% simple, ~60%
// code, ~10% resume. Expected average ≈ 140 tokens vs the prior 200,
// a ~30% reduction that compounds with the 0-token simple case
// dominating short turns.
// ---------------------------------------------------------------------------

type PromptIntent = "simple" | "code" | "resume";

/**
 * Quick keyword classifier. Heuristic-only; no LLM. Order matters:
 * resume cues are checked BEFORE code cues so "continue editing X"
 * gets resume-class treatment (which is a superset of the code
 * reminder anyway).
 */
function classifyPromptIntent(prompt: string): PromptIntent {
  const lower = prompt.toLowerCase().trim();

  // Resume signals: short prompts implying continuation. Generic
  // acks like "ok" / "next" are deliberately NOT in this list —
  // they're more often "ok thanks" / "what's next on the list" than
  // a real continuation cue. We anchor on phrases that genuinely
  // imply "pick up where we left off".
  const resumeSignals = [
    "continue",
    "resume",
    "where was i",
    "where were we",
    "carry on",
    "keep going",
    "proceed",
  ];
  if (lower.length <= 40 && resumeSignals.some((s) => lower === s || lower.startsWith(`${s} `) || lower.startsWith(`${s},`))) {
    return "resume";
  }

  // Code signals: any token from the broad code-vocabulary list. We
  // keep this list narrow enough that "code" doesn't swallow casual
  // chat but wide enough that real engineering questions hit it.
  //
  // Audit fix (2026-05-06 multi-agent fan-out, super-debugger): the
  // original list omitted history-investigation cues ("why",
  // "decision", "history", "last time"). A prompt like "why was
  // this decision made last time?" is exactly the case where
  // mneme_recall should fire, but the classifier was returning
  // "simple" and short-circuiting the reminder entirely. The
  // failing test "puts mneme_recall first for why/history prompts"
  // surfaced this — it asserted the bullet list contained
  // mneme_recall, but additionalContext was empty. Same root cause
  // for "still returns 3 tools for a short generic prompt": after
  // Item #119 introduced the simple/code/resume tiers, the test's
  // "hello" prompt correctly maps to "simple" and the test was
  // updated; the why/history-prompt test is fixed by adding the
  // missing classifier signals here so mneme_recall has a path to
  // surface.
  const codeSignals = [
    "function",
    "method",
    "class",
    "module",
    "package",
    "interface",
    "type ",
    "struct",
    "enum",
    "trait",
    "impl",
    "import",
    "export",
    "edit",
    " write ",
    "rewrite",
    "implement",
    "refactor",
    "rename",
    "delete",
    "fix",
    " bug",
    "debug",
    "trace",
    "caller",
    "callee",
    "callers",
    "callees",
    "callsite",
    "where is",
    "who calls",
    "find ",
    "search",
    "lookup",
    "blast",
    "audit",
    "test",
    "compile",
    "build",
    "deploy",
    "commit",
    "push",
    "merge",
    "rebase",
    // History / decision investigation: prompts asking WHY a thing
    // was done, or what was decided / discussed / tried previously.
    // These are exactly the cases where mneme_recall is the right
    // answer; without these signals the classifier skipped them.
    "why",
    "decision",
    "history",
    "previous",
    "last time",
    "remember",
    "recall",
    ".ts",
    ".tsx",
    ".js",
    ".jsx",
    ".rs",
    ".py",
    ".go",
    ".java",
    ".cpp",
    ".c ",
    ".cs",
    ".rb",
    "src/",
    "/src",
  ];
  if (codeSignals.some((s) => lower.includes(s))) {
    return "code";
  }

  return "simple";
}

// ---------------------------------------------------------------------------
// Prompt → tool relevance mapping
// ---------------------------------------------------------------------------

/**
 * A simple keyword → tool recommendation table. This is deliberately
 * heuristic and fast — we are not doing LLM inference at hook time. The
 * goal is to surface the 3 most likely-useful mneme tools given the prompt
 * text, so the AI has an immediate reminder in context.
 */
interface ToolRecommendation {
  readonly name: string;
  readonly why: string;
  readonly keywords: ReadonlyArray<string>;
  /** Higher = shown first when multiple recommendations score the same. */
  readonly priority: number;
}

const TOOL_RECOMMENDATIONS: ReadonlyArray<ToolRecommendation> = [
  {
    name: "mcp__mneme__blast_radius",
    why: "See what else breaks when you change this file before touching it.",
    keywords: ["edit", "change", "modify", "update", "fix", "refactor", "delete", "rename", "move", "write"],
    priority: 10,
  },
  {
    name: "mcp__mneme__mneme_recall",
    why: "Search mneme's memory for prior decisions, bugs, and context on this topic.",
    keywords: ["why", "decision", "history", "previous", "last time", "remember", "context", "background", "before", "recall"],
    priority: 9,
  },
  {
    name: "mcp__mneme__file_intent",
    why: "Understand the purpose and ownership of a file before reading or editing it.",
    keywords: ["what does", "what is", "purpose", "owned by", "responsible", "understand", "explain", "describe", "overview", "file"],
    priority: 8,
  },
  {
    name: "mcp__mneme__find_references",
    why: "Find all callers and usages before renaming or deleting a symbol.",
    keywords: ["rename", "symbol", "function", "called", "usages", "references", "callers", "where is", "who calls"],
    priority: 7,
  },
  {
    name: "mcp__mneme__call_graph",
    why: "Map the full call chain for a function — essential before deep refactors.",
    keywords: ["call", "chain", "callees", "callers", "trace", "flow", "dependency chain", "how does it reach"],
    priority: 6,
  },
  {
    name: "mcp__mneme__architecture_overview",
    why: "Get the high-level community map before adding a new feature or module.",
    keywords: ["architecture", "structure", "design", "new feature", "add module", "layer", "system", "component", "where should i"],
    priority: 5,
  },
  {
    name: "mcp__mneme__audit",
    why: "Run the drift + security scanner to catch violations before committing.",
    keywords: ["commit", "push", "deploy", "audit", "security", "lint", "check", "validate", "review", "scan"],
    priority: 5,
  },
  {
    name: "mcp__mneme__step_status",
    why: "Resume tracking the current step — never lose your place after a compaction.",
    keywords: ["step", "task", "todo", "resume", "continue", "where was i", "next", "current", "status", "progress"],
    priority: 4,
  },
  {
    name: "mcp__mneme__mneme_resume",
    why: "Get a full session brief — decisions, open questions, timeline — before diving in.",
    keywords: ["session", "catch up", "brief", "summary", "what happened", "where were we", "onboard"],
    priority: 4,
  },
  {
    name: "mcp__mneme__recall_concept",
    why: "Search semantically across code + docs + decisions for a concept.",
    keywords: ["concept", "search", "find", "semantic", "related", "similar", "about", "lookup"],
    priority: 3,
  },
] as const;

/**
 * Score a tool against a prompt. Returns a non-negative integer: higher means
 * more relevant. Zero means the tool did not match at all.
 */
function scoreToolForPrompt(tool: ToolRecommendation, promptLower: string): number {
  let score = 0;
  for (const kw of tool.keywords) {
    if (promptLower.includes(kw)) {
      score += 1;
    }
  }
  return score > 0 ? score + tool.priority : 0;
}

/**
 * Pick the top N most relevant tools for the given prompt text.
 * Always returns at least `min` tools even if no keywords match, by falling
 * back to priority order.
 */
function pickTopTools(prompt: string, n: number): ReadonlyArray<ToolRecommendation> {
  const lower = prompt.toLowerCase();

  const scored = TOOL_RECOMMENDATIONS.map((t) => ({
    tool: t,
    score: scoreToolForPrompt(t, lower),
  }));

  // Sort by score desc, then priority desc as a tiebreaker.
  scored.sort((a, b) => b.score - a.score || b.tool.priority - a.tool.priority);

  return scored.slice(0, n).map((s) => s.tool);
}

// ---------------------------------------------------------------------------
// Hook entry point
// ---------------------------------------------------------------------------

export interface UserPromptSubmitArgs {
  prompt: string;
  sessionId: string;
  cwd: string;
}

export interface UserPromptSubmitOutput {
  hook_specific: {
    additionalContext: string;
  };
}

/**
 * Build a reminder block for the AI host. The block contains:
 *   - Top 3 recommended mneme tools for this prompt.
 *   - Trespass log: Grep/Read calls this session that bypassed mneme.
 *
 * Returns an empty additionalContext (silent no-op) if the config flag
 * `inject_user_prompt_reminder` is false, or if any error occurs.
 */
export async function runUserPromptSubmit(
  args: UserPromptSubmitArgs,
): Promise<UserPromptSubmitOutput> {
  const empty: UserPromptSubmitOutput = {
    hook_specific: { additionalContext: "" },
  };

  try {
    const cfg = getHooksConfig();
    if (!cfg.inject_user_prompt_reminder) {
      return empty;
    }

    // Item #119 (2026-05-05): three-tier intent classification.
    // Simple acks pay zero tokens. Code questions get the light
    // block (top-3 tools, trespass log only if real trespasses).
    // Resume cues get the heavy block (full reminder + cue to call
    // mneme_resume / step_status — the headline compaction-recovery
    // story).
    const intent = classifyPromptIntent(args.prompt);

    if (intent === "simple") {
      // No code-related work in flight. Stay quiet.
      return empty;
    }

    const [topTools, trespasses] = await Promise.all([
      Promise.resolve(pickTopTools(args.prompt, 3)),
      getSessionTrespasses(args.sessionId, 5),
    ]);

    const additionalContext =
      intent === "resume"
        ? buildReminderBlock(topTools, trespasses, /* heavy */ true)
        : buildReminderBlock(topTools, trespasses, /* heavy */ false);

    return {
      hook_specific: { additionalContext },
    };
  } catch (err) {
    console.error("[mneme-mcp] userprompt-submit hook failed:", errMsg(err));
    return empty;
  }
}

/** Exposed for testing — see __tests__/userprompt-submit.test.ts. */
export { classifyPromptIntent };

// ---------------------------------------------------------------------------
// Block builder
// ---------------------------------------------------------------------------

/**
 * Compose the XML-tagged reminder block injected into the turn context.
 * Kept compact: the AI host already has the full session primer from
 * session_prime.ts. This block is a targeted nudge, not a full briefing.
 */
function buildReminderBlock(
  tools: ReadonlyArray<ToolRecommendation>,
  trespasses: ReadonlyArray<{ tool: string; path: string; calledAt: string }>,
  heavy: boolean,
): string {
  const lines: string[] = [];

  lines.push("<mneme-self-ping>");

  if (heavy) {
    // Item #119 (2026-05-05): heavy block fires on resume cues —
    // the headline compaction-recovery story. We assume the prior
    // turn's context may have been wiped and re-prime the AI with
    // the resume tools first, then the regular reminder.
    lines.push("Resume cue detected. Likely post-compact — these tools rebuild context fast:");
    lines.push("  • mcp__mneme__mneme_resume   — full session brief (decisions + open questions + timeline)");
    lines.push("  • mcp__mneme__step_status    — current step + acceptance criteria");
    lines.push("  • mcp__mneme__step_show      — last completed step's output");
    lines.push("");
    lines.push("If state was preserved across the compact, the step ledger has it.");
    lines.push("");
  }

  lines.push("IMPORTANT: Use mneme MCP tools BEFORE grep/read/bash for code exploration.");
  lines.push("");
  lines.push("Top 3 mneme tools for this prompt:");
  for (const t of tools) {
    lines.push(`  • ${t.name}`);
    lines.push(`    Why: ${t.why}`);
  }

  // Trespass log only emitted when real trespasses exist (light) OR
  // unconditionally during resume (heavy — the post-compact AI may
  // not even remember the trespasses happened).
  if (trespasses.length > 0) {
    lines.push("");
    lines.push("Trespass log — grep/read calls this session that skipped mneme:");
    for (const tr of trespasses) {
      const ago = formatAgo(tr.calledAt);
      lines.push(`  • ${tr.tool}("${tr.path}") — ${ago} ago, no prior mneme call`);
    }
    lines.push("Next time, call mcp__mneme__mneme_recall or blast_radius FIRST.");
  }

  lines.push("</mneme-self-ping>");

  return lines.join("\n");
}

function formatAgo(isoTimestamp: string): string {
  try {
    const diffMs = Date.now() - new Date(isoTimestamp).getTime();
    const secs = Math.floor(diffMs / 1000);
    if (secs < 60) return `${secs}s`;
    const mins = Math.floor(secs / 60);
    if (mins < 60) return `${mins}m`;
    return `${Math.floor(mins / 60)}h`;
  } catch {
    return "?";
  }
}

File: cli/src/commands/userprompt_submit.rs
//! `mneme userprompt-submit` — Wave 1E Layer 1 hook entry.
//!
//! Claude Code's settings.json registers this hook to fire on every
//! UserPromptSubmit (alongside the existing `mneme inject` hook from
//! v0.3.x). It injects a self-ping reminder block listing the top 3
//! mneme tools relevant to the user's prompt, plus an optional
//! "trespass log" of grep/read calls in this session that bypassed
//! mneme.
//!
//! ## BUG-NEW-Q fix (2026-05-05)
//!
//! HOOK_SPECS in `cli/src/platforms/claude_code.rs` registered this
//! command in Wave 1E but the Rust CLI subcommand was never wired up.
//! Every UserPromptSubmit fired `mneme.exe userprompt-submit` →
//! clap-errored with "unrecognized subcommand" → silent failure of
//! self-ping enforcement (and a brief stderr-with-non-zero-exit
//! that contributed to BUG-NEW-H's "terminal flash" report).
//!
//! This file ships the missing subcommand. The classifier mirrors
//! `mcp/src/hooks/userprompt-submit.ts` (Item #119): three tiers —
//! simple / code / resume — with simple paying zero tokens, code
//! getting the light reminder, and resume getting the heavy
//! compaction-recovery cue.
//!
//! Hook output protocol (Claude Code spec):
//!   `{ "hook_specific": { "additionalContext": "<block>" } }` on stdout.
//!   Exit 0 always — fail-open.
//!
//! Trespass-log integration is intentionally OMITTED from this Rust
//! port. The MCP TS hook reads `tool_cache.db` for that signal; doing
//! it from Rust would mean adding a sql.js dep + IPC client just for
//! the hook. v0.4.1 follow-up. The classifier alone is the
//! load-bearing piece for the token-savings story.

use clap::Args;
use serde::Deserialize;
use serde_json::json;
use std::io::{self, Read};

use crate::error::CliResult;

/// CLI args for `mneme userprompt-submit`. All optional — payload comes
/// from stdin per Claude Code's hook contract.
#[derive(Debug, Args, Default)]
pub struct UserPromptSubmitArgs {
    /// Override stdin's `prompt` field. Mostly for manual testing.
    #[arg(long)]
    pub prompt: Option<String>,
}

/// Subset of Claude Code's hook payload we actually consume.
#[derive(Debug, Default, Deserialize)]
struct StdinPayload {
    /// User's typed prompt for this turn. Empty when the hook is
    /// invoked outside a real Claude Code turn (TTY tests, manual
    /// shell invocations).
    #[serde(default)]
    prompt: String,
}

/// Three-tier classification mirroring `mcp/src/hooks/userprompt-submit.ts`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PromptIntent {
    /// Simple ack / chat. Inject nothing.
    Simple,
    /// Code-related question. Inject the light reminder.
    Code,
    /// Continuation cue. Inject the heavy compaction-recovery block.
    Resume,
}

/// Public for unit tests and parity with the TS classifier.
pub fn classify_prompt_intent(prompt: &str) -> PromptIntent {
    let lower = prompt.to_lowercase();
    let lower = lower.trim();

    // Resume signals — short prompts implying continuation. Generic
    // acks ("ok", "next", "go ahead") are deliberately excluded —
    // they're more often "ok thanks" / "what's next on the list" than
    // genuine continuation cues. Length gate (≤40 chars) prevents
    // "continue editing the spawn function" from being misclassified.
    const RESUME_SIGNALS: &[&str] = &[
        "continue",
        "resume",
        "where was i",
        "where were we",
        "carry on",
        "keep going",
        "proceed",
    ];
    if lower.len() <= 40 {
        for sig in RESUME_SIGNALS {
            if lower == *sig
                || lower.starts_with(&format!("{sig} "))
                || lower.starts_with(&format!("{sig},"))
            {
                return PromptIntent::Resume;
            }
        }
    }

    // Code signals — broad enough to catch real engineering questions,
    // narrow enough not to swallow casual chat. Order doesn't matter;
    // any single hit promotes from "simple" to "code".
    const CODE_SIGNALS: &[&str] = &[
        "function",
        "method",
        "class",
        "module",
        "package",
        "interface",
        "type ",
        "struct",
        "enum",
        "trait",
        "impl",
        "import",
        "export",
        "edit",
        " write ",
        "rewrite",
        "implement",
        "refactor",
        "rename",
        "delete",
        "fix",
        " bug",
        "debug",
        "trace",
        "caller",
        "callee",
        "callers",
        "callees",
        "callsite",
        "where is",
        "who calls",
        "find ",
        "search",
        "lookup",
        "blast",
        "audit",
        "test",
        "compile",
        "build",
        "deploy",
        "commit",
        "push",
        "merge",
        "rebase",
        ".ts",
        ".tsx",
        ".js",
        ".jsx",
        ".rs",
        ".py",
        ".go",
        ".java",
        ".cpp",
        ".c ",
        ".cs",
        ".rb",
        "src/",
        "/src",
    ];
    for sig in CODE_SIGNALS {
        if lower.contains(sig) {
            return PromptIntent::Code;
        }
    }

    PromptIntent::Simple
}

/// One recommended tool entry. Same shape as the TS hook's
/// `ToolRecommendation` so the prompt the AI sees stays identical
/// across the two hook surfaces.
struct ToolRec {
    name: &'static str,
    why: &'static str,
    keywords: &'static [&'static str],
    priority: u8,
}

/// Top-N picker. Keyword overlap + priority. Always returns N items
/// even on a no-keyword prompt by falling back to priority order.
const TOOL_RECS: &[ToolRec] = &[
    ToolRec {
        name: "mcp__mneme__blast_radius",
        why: "See what else breaks when you change this file before touching it.",
        keywords: &[
            "edit", "change", "modify", "update", "fix", "refactor", "delete", "rename", "move",
            "write",
        ],
        priority: 10,
    },
    ToolRec {
        name: "mcp__mneme__mneme_recall",
        why: "Search mneme's memory for prior decisions, bugs, and context on this topic.",
        keywords: &[
            "why",
            "decision",
            "history",
            "previous",
            "last time",
            "remember",
            "context",
            "background",
            "before",
            "recall",
        ],
        priority: 9,
    },
    ToolRec {
        name: "mcp__mneme__file_intent",
        why: "Understand the purpose and ownership of a file before reading or editing it.",
        keywords: &[
            "what does",
            "what is",
            "purpose",
            "owned by",
            "responsible",
            "understand",
            "explain",
            "describe",
            "overview",
            "file",
        ],
        priority: 8,
    },
    ToolRec {
        name: "mcp__mneme__find_references",
        why: "Find all callers and usages before renaming or deleting a symbol.",
        keywords: &[
            "rename",
            "symbol",
            "function",
            "called",
            "usages",
            "references",
            "callers",
            "where is",
            "who calls",
        ],
        priority: 7,
    },
    ToolRec {
        name: "mcp__mneme__call_graph",
        why: "Map the full call chain for a function — essential before deep refactors.",
        keywords: &[
            "call",
            "chain",
            "callees",
            "callers",
            "trace",
            "flow",
            "dependency chain",
            "how does it reach",
        ],
        priority: 6,
    },
    ToolRec {
        name: "mcp__mneme__architecture_overview",
        why: "Get the high-level community map before adding a new feature or module.",
        keywords: &[
            "architecture",
            "structure",
            "design",
            "new feature",
            "add module",
            "layer",
            "system",
            "component",
            "where should i",
        ],
        priority: 5,
    },
    ToolRec {
        name: "mcp__mneme__audit",
        why: "Run the drift + security scanner to catch violations before committing.",
        keywords: &[
            "commit", "push", "deploy", "audit", "security", "lint", "check", "validate", "review",
            "scan",
        ],
        priority: 5,
    },
    ToolRec {
        name: "mcp__mneme__step_status",
        why: "Resume tracking the current step — never lose your place after a compaction.",
        keywords: &[
            "step",
            "task",
            "todo",
            "resume",
            "continue",
            "where was i",
            "next",
            "current",
            "status",
            "progress",
        ],
        priority: 4,
    },
    ToolRec {
        name: "mcp__mneme__mneme_resume",
        why: "Get a full session brief — decisions, open questions, timeline — before diving in.",
        keywords: &[
            "session",
            "catch up",
            "brief",
            "summary",
            "what happened",
            "where were we",
            "onboard",
        ],
        priority: 4,
    },
    ToolRec {
        name: "mcp__mneme__recall_concept",
        why: "Search semantically across code + docs + decisions for a concept.",
        keywords: &[
            "concept", "search", "find", "semantic", "related", "similar", "about", "lookup",
        ],
        priority: 3,
    },
];

fn pick_top_tools(prompt: &str, n: usize) -> Vec<&'static ToolRec> {
    let lower = prompt.to_lowercase();
    let mut scored: Vec<(u32, &'static ToolRec)> = TOOL_RECS
        .iter()
        .map(|t| {
            let mut s = 0u32;
            for kw in t.keywords {
                if lower.contains(kw) {
                    s += 1;
                }
            }
            let total = if s > 0 { s + u32::from(t.priority) } else { 0 };
            (total, t)
        })
        .collect();
    scored.sort_by(|a, b| b.0.cmp(&a.0).then_with(|| b.1.priority.cmp(&a.1.priority)));
    scored.into_iter().take(n).map(|(_, t)| t).collect()
}

fn build_reminder_block(tools: &[&ToolRec], heavy: bool) -> String {
    let mut s = String::with_capacity(512);
    s.push_str("<mneme-self-ping>\n");
    if heavy {
        s.push_str(
            "Resume cue detected. Likely post-compact — these tools rebuild context fast:\n",
        );
        s.push_str(
            "  • mcp__mneme__mneme_resume   — full session brief (decisions + open questions + timeline)\n",
        );
        s.push_str("  • mcp__mneme__step_status    — current step + acceptance criteria\n");
        s.push_str("  • mcp__mneme__step_show      — last completed step's output\n");
        s.push_str("\nIf state was preserved across the compact, the step ledger has it.\n\n");
    }
    s.push_str("IMPORTANT: Use mneme MCP tools BEFORE grep/read/bash for code exploration.\n\n");
    s.push_str("Top 3 mneme tools for this prompt:\n");
    for t in tools {
        s.push_str(&format!("  • {}\n    Why: {}\n", t.name, t.why));
    }
    s.push_str("</mneme-self-ping>");
    s
}

/// Entry point — wired into `cli/src/main.rs`.
///
/// Fail-open: any error returns an empty additionalContext + exit 0.
/// We MUST NOT block the user's prompt because of a hook bug.
pub async fn run(args: UserPromptSubmitArgs) -> CliResult<()> {
    // Read STDIN payload. Empty / malformed → empty default.
    let prompt = match args.prompt {
        Some(p) => p,
        None => read_stdin_prompt().unwrap_or_default(),
    };

    let intent = classify_prompt_intent(&prompt);

    let additional_context = match intent {
        PromptIntent::Simple => String::new(),
        PromptIntent::Code => {
            let tools = pick_top_tools(&prompt, 3);
            build_reminder_block(&tools, false)
        }
        PromptIntent::Resume => {
            let tools = pick_top_tools(&prompt, 3);
            build_reminder_block(&tools, true)
        }
    };

    let out = json!({
        "hook_specific": { "additionalContext": additional_context },
    });
    println!("{}", out);
    Ok(())
}

/// Best-effort stdin read. Returns None on any IO failure (so the
/// caller defaults to an empty prompt and a Simple intent).
fn read_stdin_prompt() -> Option<String> {
    let mut buf = String::new();
    if io::stdin().read_to_string(&mut buf).is_err() {
        return None;
    }
    let payload: StdinPayload = serde_json::from_str(&buf).ok().unwrap_or_default();
    Some(payload.prompt)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_acks_classify_simple() {
        for p in [
            "ok thanks",
            "thanks!",
            "got it",
            "great",
            "nice",
            "what next",
            "hello",
            "how are you",
        ] {
            assert_eq!(classify_prompt_intent(p), PromptIntent::Simple, "{p}");
        }
    }

    #[test]
    fn code_questions_classify_code() {
        for p in [
            "find all callers of WorkerPool::spawn",
            "rewrite the audit function",
            "edit cli/src/main.rs",
            "what's the type of EdgeKind?",
            "show me the bug in build.rs",
            "blast radius of changing PathManager",
            "fix the typescript file at vision/src/api.ts",
        ] {
            assert_eq!(classify_prompt_intent(p), PromptIntent::Code, "{p}");
        }
    }

    #[test]
    fn resume_cues_classify_resume() {
        for p in [
            "continue",
            "where was i",
            "resume",
            "carry on",
            "keep going",
            "proceed",
            "continue, please",
            "resume from where we left off",
        ] {
            assert_eq!(classify_prompt_intent(p), PromptIntent::Resume, "{p}");
        }
    }

    #[test]
    fn long_continue_prompt_is_code_not_resume() {
        // Length gate keeps "continue editing X.rs" out of resume.
        let p =
            "continue editing the WorkerPool spawn function in manager.rs and add error handling";
        assert_eq!(classify_prompt_intent(p), PromptIntent::Code);
    }

    #[test]
    fn pick_top_tools_returns_n_even_with_no_keyword_match() {
        let tools = pick_top_tools("hello", 3);
        assert_eq!(tools.len(), 3);
    }

    #[test]
    fn build_reminder_block_heavy_includes_resume_lead_in() {
        let tools = pick_top_tools("continue", 3);
        let block = build_reminder_block(&tools, true);
        assert!(block.contains("Resume cue detected"));
        assert!(block.contains("mneme_resume"));
    }

    #[test]
    fn build_reminder_block_light_omits_resume_lead_in() {
        let tools = pick_top_tools("fix the bug in foo.rs", 3);
        let block = build_reminder_block(&tools, false);
        assert!(!block.contains("Resume cue detected"));
        assert!(block.contains("Top 3 mneme tools"));
    }
}
